shared/src/econ/mod.rs

```diff
@@ -1,7 +1,345 @@
-// Copyright 2019-2022 ChainSafe Systems
-// SPDX-License-Identifier: Apache-2.0, MIT
+use std::cmp::Ordering;
+use std::fmt;
+use std::ops::{Add, Mul, Neg, Sub};
 
 use num_bigint::BigInt;
+use num_integer::Integer;
+use num_traits::{Signed, Zero};
+use serde::{Deserialize, Serialize, Serializer};
 
-/// Token type to be used within the VM.
-pub type TokenAmount = BigInt;
+use crate::bigint::bigint_ser;
+
+/// A quantity of native tokens.
+/// A token amount is an integer, but has a human interpretation as a value with
+/// 18 decimal places.
+/// This is a new-type in order to prevent accidental conversion from other BigInts.
+/// From/Into BigInt is missing by design.
+#[derive(Clone, PartialEq, Eq, Hash)]
+pub struct TokenAmount {
+    atto: BigInt,
+}
+
+// This type doesn't implement all the numeric traits (Num, Signed, etc),
+// opting for a minimal useful set. Others can be added if needed.
+impl TokenAmount {
+    /// The logical number of decimal places of a token unit.
+    pub const DECIMALS: usize = 18;
+
+    /// The logical precision of a token unit.
+    pub const PRECISION: u64 = 10u64.pow(Self::DECIMALS as u32);
+
+    /// Creates a token amount from a quantity of indivisible units  (10^-18 whole units).
+    pub fn from_atto(atto: BigInt) -> Self {
+        Self { atto }
+    }
+
+    /// Creates a token amount from a quantity of indivisible units (10^-18 whole units).
+    pub fn from_atto_u128(atto: u128) -> Self {
+        Self {
+            atto: BigInt::from(atto),
+        }
+    }
+
+    /// Creates a token amount from a quantity of indivisible units (10^-18 whole units).
+    pub fn from_atto_u64(atto: u64) -> Self {
+        Self {
+            atto: BigInt::from(atto),
+        }
+    }
+
+    /// Creates a token amount from a quantity of whole units (10^18 indivisible units).
+    pub fn from_whole(tokens: i64) -> Self {
+        Self {
+            atto: BigInt::from(tokens) * Self::PRECISION,
+        }
```

Comment:
```suggestion
        Self::from_atto((tokens as i128) * (Self::PRECISION as i128))
```

Slightly faster, and it re-uses other methods.